### PR TITLE
Miscellaneous suggestions for improvement and contributions, typo correction

### PR DESCRIPTION
**Pull request by vboussange**
Hey there, 
This review and its philosophy are brilliant ideas. This piece of work is so much required by the (at least my) community, and I believe it can have a substantial impact.

Here are a few remarks, that I have also included in the manuscript
- [ ] I feel like how the application of SA and DP is relevant to the different scientific domains mentioned could be an important part of the review. We could put more emphasis on its importance in the abstract, plain language summary, and introduction.
- [ ] The introduction structure is to be improved. The context and aim of the paper should be clearly stated, and the structure of the paper should be described.
- [ ] Although sensitivity analysis is the initial focus of the paper, I think that it would be very relevant to have an additional section "Statistical foundations of inverse modelling", to better understand how sensitivity analysis connects to maximum likelihood and maximum a priori estimation, as well as Bayesian inference with Markov chain Monte Carlo. Right now, the nature of "Empirical loss functions." is somewhat assumed completetely different than that of "Likelihood profiles.", while I believe that understanding the link is important, especially in SciML.
   - On this topic, [Schartau et al.]https://bg.copernicus.org/articles/14/1647/2017/(), [Schneider et al.](https://onlinelibrary.wiley.com/doi/10.1002/2017GL076101) and [Boquet et al.](https://npg.copernicus.org/articles/26/143/2019/) provide a nice template. I would be happy to lead this section.

- [ ] Considering these remarks, I suggest the following revised structure:
- Introduction
- Inverse modelling in science
  - Inverse modelling
  - Challenges
  - Domain specific applications
- Statistical foundations of inverse modelling
  - Models and data 
  - Maximum likelihood estimation and loss function 
  - Bayesian inference with MCMC
  - Model sensitivity to parameters
  - Diagnosis of the solution
- Sensitivity analysis methods
  - ...
- Numerical implementation 
  - ...
- Recommendations

- [ ] In the recommendation section, for inverse problems, practitioners would surely benefit from recommendations on which type of optimizer to use. An experiment which shows how e.g. SDG, Adam, Radam, BFGS and LBFGS perform on the minimzation of a loss involving different model types, in terms of convergence vs computational complexity, could help to discuss this. I would be happy to provide such an experiment. 
- [ ] I would be happy to lead the subsection Biology and Ecology.

### CODE REVIEW

File: tex/sections/software.tex [hunk 1]
```diff
@@ -81,18 +82,18 @@ \subsection{Direct methods}
...
 This difference between the methods also makes the complex step method sometimes more efficient than both finite differences and AD \cite{Lantoine_Russell_Dargent_2012}, an effect that can be counterbalanced by the number of extra unnecessary operations that complex arithmetic requires (see last column of \eqref{eq:AD-complex-comparision}) \cite{Martins_Sturdza_Alonso_2003_complex_differentiation}.
-It is also important to remark that many modern software already have support for complex number arithmetic, making complex step differentiation very easy to implement.
+It is also important to remark \todo{Repetition} that many modern software already have support for complex number arithmetic, making complex step differentiation very easy to implement.
 
 % \subsubsection{Reverse-mode AD}
 
-The libraries \texttt{ReverseDiff.jl} and \texttt{Zygote.jl} use callbacks to compute gradients. 
+The libraries \texttt{ReverseDiff.jl} and \texttt{Zygote.jl} use callbacks to compute gradients. \todo{We need to define what are callbacks}
```
Comment (facusapienza21) [inline tex/sections/software.tex:91]: I think this has been done in the new version.

File: tex/sections/software.tex [hunk 2]
```diff
@@ -81,18 +82,18 @@ \subsection{Direct methods}
     = 
     \, \cos(x^2) (2x).
 \end{equation}
-The stepsize dependence of the complex step differentiation method makes it resemble more to finite differences than AD with dual numbers. 
+The dependence of the complex step differentiation method on the step size gives it a closer resemblance to finite difference methods than to automatic differentiation (AD) using dual numbers.
```
Comment (facusapienza21) [inline tex/sections/software.tex:85]: Perfect

File: tex/sections/software.tex [hunk 3]
```diff
@@ -81,18 +82,18 @@ \subsection{Direct methods}
...
 This difference between the methods also makes the complex step method sometimes more efficient than both finite differences and AD \cite{Lantoine_Russell_Dargent_2012}, an effect that can be counterbalanced by the number of extra unnecessary operations that complex arithmetic requires (see last column of \eqref{eq:AD-complex-comparision}) \cite{Martins_Sturdza_Alonso_2003_complex_differentiation}.
-It is also important to remark that many modern software already have support for complex number arithmetic, making complex step differentiation very easy to implement.
+It is also important to remark \todo{Repetition} that many modern software already have support for complex number arithmetic, making complex step differentiation very easy to implement.
```
Comment (facusapienza21) [inline tex/sections/software.tex:87]: Perfect

File: tex/sections/software.tex [hunk 4]
```diff
@@ -101,19 +102,19 @@ \subsubsection{Further remarks}
...
 
 It is important to remark that the underlying machinery of all solvers relies on solvers for linear systems of equations, which can be solved in dense, band (sparse), and Krylow mode. 
 % This implies that methods based on numerical solvers are, in principle, more difficult to implement but also more efficient in computing gradients for complex differential equations. 
 Another important consideration is that all these methods have subroutines to compute the VJPs involved in the sensitivity and adjoint equations. 
-This calculation is carried out by another sensitivity method (finite differences, AD) and this also plays a central role at the moment of analyzing the accuracy and stability of the adjoint method. 
+This calculation is carried out by another sensitivity method (finite differences, AD) which plays a central role when analyzing the accuracy and stability of the adjoint method. 
 
-\subsubsection{Sensitivity equation}
+\subsubsection{Sensitivity equation}\todo{seems like a repetition}
```
Comment (facusapienza21) [inline tex/sections/software.tex:117]: This refers to the method of sensitivity equations, all the names are a little bit similar XD

File: tex/sections/software.tex [hunk 5]
```diff
@@ -101,19 +102,19 @@ \subsubsection{Further remarks}
...
+These include \texttt{CVODES} within \texttt{SUNDIALS} in C \cite{serban2005cvodes, SUNDIALS-hindmarsh2005sundials}; \texttt{ODESSA} \cite{ODESSA} and \texttt{FATODE} (discrete adjoints) \cite{FATODE2014} both in Fortram; \texttt{SciMLSensitivity.jl} in Julia \cite{rackauckas2020universal}; \texttt{Dolfin-adjoint} based on the \texttt{FEniCS} Project \cite{dolfin2013, dolfin2018}. 
 
 It is important to remark that the underlying machinery of all solvers relies on solvers for linear systems of equations, which can be solved in dense, band (sparse), and Krylow mode. 
 % This implies that methods based on numerical solvers are, in principle, more difficult to implement but also more efficient in computing gradients for complex differential equations. 
 Another important consideration is that all these methods have subroutines to compute the VJPs involved in the sensitivity and adjoint equations. 
-This calculation is carried out by another sensitivity method (finite differences, AD) and this also plays a central role at the moment of analyzing the accuracy and stability of the adjoint method.
```
Comment (facusapienza21) [inline tex/sections/software.tex:114]: Perfect

File: tex/sections/software.tex [hunk 6]
```diff
@@ -81,18 +82,18 @@ \subsection{Direct methods}
...
 % \subsubsection{Reverse-mode AD}
 
-The libraries \texttt{ReverseDiff.jl} and \texttt{Zygote.jl} use callbacks to compute gradients. 
+The libraries \texttt{ReverseDiff.jl} and \texttt{Zygote.jl} use callbacks to compute gradients. \todo{We need to define what are callbacks}
 When gradients are being computed with less than $\sim 100$ parameters, the former is faster (see documentation).
 
 \cite{Baydin_Pearlmutter_Radul_Siskind_2015}
 
-Notice that the application of reverse AD on a numerical solver (without checkpointing) scales as $\mathcal O (n k)$, with $k$ the number of steps of the numerical solver. 
+Notice that the application of reverse AD on a numerical solver (without checkpointing) scales as $\mathcal O (n k)$, with $k$ the number of steps of the numerical solver. \todo{I think that we should introduce checkpointing before}
```
Comment (facusapienza21) [inline tex/sections/software.tex:96]: I agree. It is still unclear to me where is the best part to introduce this.

File: tex/sections/statistical-foundations.tex
Comment (facusapienza21) [inline tex/sections/statistical-foundations.tex:1]: I think we can add this file to the repo, but based on our conversation last time @vboussange I think is better to keep it outside the main text for now. Maybe include it in SM at some point, but I think right now it goes a little bit outside the scope of the paper.

**Review by facusapienza21 — CHANGES_REQUESTED** [review]
Thank you @vboussange for the PR! I made comments on each file, either directly approving changes or asking to make some changes. 

If possible, I will try to remove as many `\todo{}` latex commands from the text and include those as issues or commented text, to make the current version already readable for interested people (althoght we know this is still work in progress, but I know of some poeple using it for reference already).

File: tex/main.tex [hunk 7]
```diff
@@ -111,14 +116,17 @@ \subsubsection{Discrete adjoint method}
...
 \label{sec:computational-implementation}
 \input{sections/software}
 
+
 % \section{Extensions}
 
 \section{Recommendations}
 \input{sections/recommendations}
+\todo[inline]{Practitioners would surely benefit from recommendations on which type of optimizer to use, in the case of inverse modelling. 
+For instance, it is useful to know that Adam is much less computationally demanding than BFGS, and is less prone to converging to local minimum. Adam + BFGS is a nice combination to avoid local minima while obtaining a precise parameter estimation. An experiment which shows how e.g. SDG, Adam, Radam, BFGS and LBFGS perform on the minimzation of a loss involving different models, in terms of convergence vs computational complexity, could help to discuss this. I would be happy to provide such an experiment.}
```
[the remark this answers is not in the document]
Comment (vboussange) [reply tex/main.tex:129]: I have tried to include this in section 4.1. Let me know what you think.

File: tex/sections/software.tex (hunk 6 above)
Comment (vboussange) [reply tex/sections/software.tex:96]: I opened an issue.

File: tex/sections/statistical-foundations.tex
Comment (vboussange) [reply tex/sections/statistical-foundations.tex:1]: OK.

**Review by vboussange — COMMENTED** [review]
Hey @facusapienza21, I have reviewed your comments - I let you do the merge. I will address #92 by next week. Should I make a new PR for this?

**Comment by facusapienza21** [comment]
Thank you @vboussange !!! Looking great. I will review the PR in these days and make the merge. I may make some small modifications for formatting, if something pops up I will let you know. Otherwise, I will just integrate the changes. 

The current version in main is a little but outdates with respect to the one in my fork, so I will later have to merge those changes too (meaning, some things had been changed). 

If you are ok working with PR, that is great! I like this workflow. However, if you prefer to be added to the Overleaf available from my fork and make changes there, let me know! Whatever is more convenient for you.

**Comment by vboussange** [comment]
Hey @facusapienza21, I would actually prefer working with the Overleaf, if possible!

**Pull request merged**